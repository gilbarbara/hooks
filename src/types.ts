import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
