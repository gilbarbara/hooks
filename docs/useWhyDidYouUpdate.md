# useWhyDidYouUpdate

Get which props/state changes are causing a component to re-render.

## Usage

```tsx
function Component(props: any) {
  const changes = useWhyDidYouUpdate(props, { skipLog: true });
  // Or just log the changes
  // useWhyDidYouUpdate(props, 'Component');

  return <div>{!!changes && JSON.stringify(changes, null, 2)}</div>;
}
```

## Reference

```typescript
interface UseWhyDidYouUpdateOptions {
  name?: string;
  skipLog?: boolean;
}

type UseWhyDidYouUpdateResult<T> = {
  [K in keyof T]?: {
    from: any;
    to: any;
  };
};

useWhyDidYouUpdate<T extends Record<string, any>>(
  props: T,
  nameOrOptions: string | UseWhyDidYouUpdateOptions = {},
): UseWhyDidYouUpdateResult<T>;
```
