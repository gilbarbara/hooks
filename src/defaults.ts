import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
