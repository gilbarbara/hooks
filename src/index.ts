export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
