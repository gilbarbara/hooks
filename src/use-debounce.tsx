import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
