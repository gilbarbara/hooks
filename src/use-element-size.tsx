import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
