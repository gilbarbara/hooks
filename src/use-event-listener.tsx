import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
