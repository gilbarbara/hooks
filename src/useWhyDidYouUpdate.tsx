import { useEffect, useRef, useState } from 'react';

import { PlainObject } from './types';
import { isString } from './utils';

export type UseWhyDidYouUpdateResult<T> = {
  [K in keyof T]?: {
    from: any;
    to: any;
  };
};

export interface UseWhyDidYouUpdateOptions {
  name?: string;
  skipLog?: boolean;
}

export function useWhyDidYouUpdate<T extends PlainObject<any>>(
  props: T,
  nameOrOptions: string | UseWhyDidYouUpdateOptions = {},
): UseWhyDidYouUpdateResult<T> | false {
  type K = keyof T;

  const [changes, setChanges] = useState<UseWhyDidYouUpdateResult<T>>({});
  const previousProps = useRef<T>();

  const { name, skipLog = false } = isString(nameOrOptions)
    ? { name: nameOrOptions }
    : nameOrOptions;

  useEffect(() => {
    const { current } = previousProps;

    setChanges({});

    if (current) {
      // Get all keys from previous and current props
      const allKeys = Object.keys({ ...current, ...props }) as K[];

      // Use this object to keep track of changed props
      const changesObject: UseWhyDidYouUpdateResult<T> = {};

      // Iterate through keys
      allKeys.forEach(key => {
        if (current[key] !== props[key]) {
          changesObject[key] = {
            from: current[key],
            to: props[key],
          };
        }
      });

      // If changesObj not empty then output to console
      if (Object.keys(changesObject).length) {
        setChanges(changesObject);

        if (!skipLog) {
          const nameToken = name ? `: ${name}` : '';

          console.log(`[why-did-you-update${nameToken}]`, changesObject); // eslint-disable-line no-console
        }
      }
    }

    // Finally update previousProps with current props for next hook call
    previousProps.current = props;
  }, [name, props, skipLog]);

  return Object.keys(changes).length ? changes : false;
}
