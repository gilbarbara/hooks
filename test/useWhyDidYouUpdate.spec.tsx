import * as React from 'react';
import { render } from '@testing-library/react';

import { useWhyDidYouUpdate } from '../src/useWhyDidYouUpdate';

let name: string;
let skipLog: boolean;

function Component(props: any) {
  const changes = useWhyDidYouUpdate(props, { name, skipLog });

  return <div>{!!changes && JSON.stringify(changes, null, 2)}</div>;
}

describe('useWhyDidYouUpdate', () => {
  const { log } = console;

  beforeAll(() => {
    console.log = jest.fn();
  });

  afterAll(() => {
    console.log = log;
  });

  afterEach(() => {
    // @ts-ignore
    console.log.mockReset();
  });

  it('should show the changes', () => {
    skipLog = true;
    const { container, rerender } = render(<Component version={1} />);

    expect(container.firstChild).toMatchSnapshot();

    rerender(<Component version={2} />);
    expect(container.firstChild).toMatchSnapshot();

    rerender(<Component version={2} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should log the changes', () => {
    skipLog = false;
    const { rerender } = render(<Component version={1} />);

    expect(console.log).toHaveBeenCalledTimes(0);

    rerender(<Component version={2} />);

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenLastCalledWith('[why-did-you-update]', {
      version: { from: 1, to: 2 },
    });

    rerender(<Component version={2} />);
    expect(console.log).toHaveBeenCalledTimes(1);
  });

  it('should log the changes with `name`', () => {
    name = 'Component';
    skipLog = false;
    const { rerender } = render(<Component version={1} />);

    expect(console.log).toHaveBeenCalledTimes(0);

    rerender(<Component version={2} />);

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenLastCalledWith('[why-did-you-update: Component]', {
      version: { from: 1, to: 2 },
    });
  });
});
