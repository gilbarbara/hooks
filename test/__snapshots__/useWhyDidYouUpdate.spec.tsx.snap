// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useWhyDidYouUpdate > should show the changes 1`] = `
<div>
  false
</div>
`;

exports[`useWhyDidYouUpdate > should show the changes 2`] = `
<div>
  {
  "version": {
    "from": 1,
    "to": 2
  }
}
</div>
`;

exports[`useWhyDidYouUpdate > should show the changes 3`] = `
<div>
  false
</div>
`;
